scripts/prepare_doc_publication.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Prepares MarkDown documentation publication.
+
+The in-tree and auto-generated MarkDown documentation lacks necessary metadata
+(i.e., front matter) for specifying the layout, title, and others that are
+required by Jekyll for rendering the HTML. This script patches MarkDown
+files with that.
+"""
+
+import argparse
+import os
+
+
+def parse_arguments():
+    """Parses command-line options."""
+    parser = argparse.ArgumentParser(description='Processes MarkDown files for publication')
+    parser.add_argument('base_dir', metavar='PATH', type=str, help='Base documentation directory.')
+
+    args = parser.parse_args()
+    if not os.path.isdir(args.base_dir):
+        raise parser.error('expected path to a directory')
+
+    return args
+
+# A dictionary containing source file to permanent link mappings.
+#
+# By default a source file will have a permanent URL link following its
+# filename. For example, if we have docs/Foo/Bar.md, then the permanent link
+# for it would be https://google.github.io/iree/Foo/Bar. This dictionary
+# allows one to override the permanent link if necessary.
+PERMALINK_DICT = {
+    'roadmap_design.md': "DesignRoadmap",
+}
+
+# A dictionary containing source file to navigation order mappings.
+#
+# By default the rendered docs will be sort alphabetically and listed on
+# the left panel of https://google.github.io/iree website. This allows one
+# to specify an order for a specific doc.
+NAVI_ORDER_DICT = {
+    # 'Home' is 1.
+    'roadmap_design.md': 2,
+}
+
+def process_file(basedir, relpath, filename):
+    """Modifies the given file in-place with metadata and other changes to
+    make it suitable for publication."""
+
+    full_path = os.path.join(basedir, relpath, filename)
+    base_name = os.path.splitext(filename)[0]
+    with open(full_path, 'r') as f:
+        content = f.read()
+
+    # Directly return if the file already has front matter.
+    if content.startswith('---\n'):
+        return
+
+    front_matter = {}
+    # Use the default layout for everything.
+    front_matter['layout'] = 'default'
+    # Use the base filename as permanent link.
+    front_matter['permalink'] = base_name
+
+    # Organize each doc to a section matching its directory structure.
+    if relpath and relpath != '.':
+        front_matter['parent'] = relpath
+        front_matter['permalink'] = '{}/{}'.format(relpath, front_matter['permalink'])
```

Comment:
Consider f strings. We're switching bazel_to_cmake to those in https://github.com/google/iree/pull/1302. https://realpython.com/python-f-strings/

Reply:
TIL. Thanks!